### PR TITLE
OpenCL improvements

### PR DESCRIPTION
**Pull request by lukeiwanski**
- Bumps Eigen Version
 - Refactors Ops registration
 - Introduces workaround for Const Op related to the difference between
   CUDA which uses pointers and OpenCL that uses buffers/accessors
 - Extends memory types to cover DEVICE_SYCL as well
 - Introduces  GetSYCLDevice() method that returns list of supported devices
   with GPU device having the highest priority ( doesn't include blacklisted devices )
 - ::internal::Transpose -> tensorflow::internal::Transpose in order to
   avoid compilation reported error
 - re-introduces fix for bugged string replacement causing a lot of compilation
   warnings -c -> --include
 - Adds sycl_runtime to bazels ARRAY_DEPS
 - Replicates TF_CALL_GPU_PROXY_TYPES for SYCL

### CODE REVIEW

File: tensorflow/core/kernels/reverse_op.cc [hunk 1]
```diff
@@ -29,6 +29,10 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/util/work_sharder.h"
 
+#ifdef TENSORFLOW_USE_SYCL
+#include "tensorflow/core/common_runtime/sycl/sycl_util.h"
+#endif // TENSORFLOW_USE_SYCL
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/reverse_op.cc]: Why do you include sycl_util.h here ?

File: tensorflow/core/kernels/stack_ops.cc [hunk 2]
```diff
@@ -308,22 +308,20 @@ REGISTER_GPU_HOST_KERNEL(bool);
...
-#undef REGISTER_SYCL_KERNEL
 #define REGISTER_SYCL_HOST_KERNEL(type)                   \
   REGISTER_KERNEL_BUILDER(Name("StackPush")               \
                               .Device(DEVICE_SYCL)        \
                               .HostMemory("handle")       \
                               .HostMemory("elem")         \
                               .HostMemory("output")       \
                               .TypeConstraint<type>("T"), \
-                          StackPushOp<SYCLDevice>)
+                          StackPushOp<GPUDevice>)
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/stack_ops.cc]: Why not SYCLevice here ?

File: tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc [hunk 3]
```diff
@@ -169,6 +176,10 @@ static void TestHWAccelerator(bool enableHWTrace) {
   SessionOptions options;
   (*options.config.mutable_device_count())["CPU"] = 1;
   (*options.config.mutable_device_count())["GPU"] = 1;
+#ifdef TENSORFLOW_USE_SYCL
```
Comment (benoitsteiner) [inline tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc]: Why do you assign x and y to sycl devices above only to ask tensorflow to ignore the sycl device ?

Comment (lukeiwanski) [inline tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc]: The cost model for SYCL is not yet implemented ( an incorrect value for the model is returned ). Re-enabled in 6c7f9cf.

File: tensorflow/core/kernels/reverse_op.cc (hunk 1 above)
Comment (lukeiwanski) [inline tensorflow/core/kernels/reverse_op.cc]: That was caused by broken merge fixed in 6c7f9cf

File: tensorflow/core/kernels/stack_ops.cc (hunk 2 above)
Comment (lukeiwanski) [inline tensorflow/core/kernels/stack_ops.cc]: Typo, fixed in 6c7f9cf

**Pull request merged**